cpu/sam3x8e/periph/pwm.c

```diff
@@ -0,0 +1,258 @@
+ *
+ * \todo { Find out why simple assignment '=' isn't working although
+ * registers are Read only}
+ * \todo { Extend macro use }
+ */
+#define PWM_0_CHAN_PIN_SET() {                          \
+        PIOC->PIO_PDR  |= PIO_PDR_P21;                  \
+        PWM_0_PORT_CH0->PIO_ABSR |= PWM_0_PIN_CH0;      \
+        PIOC->PIO_PDR  |= PIO_PDR_P22;                  \
+        PWM_0_PORT_CH1->PIO_ABSR |= PWM_0_PIN_CH1;      \
+        PIOC->PIO_PDR  |= PIO_PDR_P23;                  \
+        PWM_0_PORT_CH2->PIO_ABSR |= PWM_0_PIN_CH2;      \
+        PIOC->PIO_PDR  |= PIO_PDR_P24;                  \
+        PWM_0_PORT_CH3->PIO_ABSR |= PWM_0_PIN_CH3;      \
+    }
+
+/*!
+ * Only handles one device with max. four channels.
+ *
+ * To ease initialisation the four channels (4-7) are controlled by a sync
+ * group. CH0 is master of sync group and inherently selected.
+ *
+ * - Selecting clock and mode.
+ * - Set period and other comparator paramters.
+ * - En/Disable.
+ *
+ * Duty-cycle update is selected individualy.
+ *
+ * \warning {Although channel 0 isn't routed to a pin, controlling th sync
+ * group means it MUST NOT used for other purposes, Channels 1-3 are still
+ * free usable}
+ * \todo { Use PWM_0_CHANNELS for selective initialisation }
+ */
+int pwm_init (pwm_t dev, pwm_mode_t mode,
+              unsigned int frequency,
+              unsigned int resolution)
+{
+
+    int32_t retval = ERR_INIT_MODE;      /* Worst/First case */
+    uint32_t pwm_clk = 0;                //!< Desired/real pwm_clock //
+    uint32_t diva = 1;                   //!< Candidate for 8bit prescaler //
+    uint32_t prea = 0;                   //!< Candidate for clock select //
+
+    Pwm* pwm;
+    PwmCh_num* chan;
+
+    /*
+     * Mode check.
+     * Involves things that shold be in "periph_conf.h".
+     * !Junk!Just wasting memory an a few cpu cycles for now. ;/
+     */
+    switch ( mode) {
+    case PWM_LEFT:
+        //!\todo
+        break;
+    case PWM_CENTER:
+        //!\todo
+        break;
+    default:
+        //retval = ERR_INIT_BWTH;
+        // Don't like :/
+        return ERR_INIT_BWTH;
+    }
+
+    /* Should check if "|log_2 frequency|+|log_2 resolution| <= 32" */
+    pwm_clk = frequency * resolution;
+
+    /*
+     * Sanity check.
+     *
+     * The pwm provides 11 prescaled clocks with (MCK/2^prea | prea=[0,10])
+     * and a divider (diva) with a denominator range [1,255] in line.
+     * With every step the prescaler is doubled and the remaining quotient
+     * is halfed and tested <=255.
+     */
+    if ( F_CPU < pwm_clk) { // Have to cut down resulting frequency.
+        frequency = F_CPU/resolution;
+    } else { // Estimate prescaler and divider.
+        diva = F_CPU / pwm_clk;
+        while ( ( prea < MCK_DIV_LB_MAX) && ( ~0xff & diva)) {
+            prea = prea + 1;
+            diva = diva >> 1;
+        }
+        frequency = F_CPU / ( ( resolution * diva) << prea);
+    }
```

Comment:
I didn't go into details here but did you compare other drivers? Your solution looks a bit complicated to me. Also I think the while-loop can be avoided.